pkg/pillar/zedcloud/deferred.go

```diff
@@ -158,69 +228,100 @@ func (ctx *DeferredContext) removeDeferred(log *base.LogObject, key string) {
+	log.Functionf("SetDeferred(%s) size %d items %d\n",
 		key, size, len(ctx.deferredItems))
 	if len(ctx.deferredItems) == 0 {
 		startTimer(log, ctx)
 	}
-	_, ok := ctx.deferredItems[key]
-	if ok {
-		log.Tracef("Replacing key %s\n", key)
-	} else {
-		log.Tracef("Adding key %s\n", key)
-	}
 	item := deferredItem{
+		added:         time.Now(),
 		buf:           buf,
 		size:          size,
 		url:           url,
 		zedcloudCtx:   zedcloudCtx,
 		bailOnHTTPErr: bailOnHTTPErr,
 	}
-	l := deferredItemList{}
+	l := deferredItemList{key: key, itemType: itemType}
 	l.list = append(l.list, item)
-	ctx.deferredItems[key] = l
+	found := false
+	ind := 0
+	var itemList deferredItemList
+	for ind, itemList = range ctx.deferredItems {
+		if itemList.key == key {
+			found = true
+			break
+		}
+	}
+	if found {
+		log.Tracef("Replacing key %s\n", key)
+		ctx.deferredItems[ind] = l
+	} else {
+		log.Tracef("Adding key %s\n", key)
+		ctx.deferredItems = append(ctx.deferredItems, l)
+	}
 }
 
 // Add to slice for this key
 func AddDeferred(zedcloudCtx *ZedCloudContext, key string, buf *bytes.Buffer,
-	size int64, url string, bailOnHTTPErr bool) {
+	size int64, url string, bailOnHTTPErr bool, itemType interface{}) {
 
-	zedcloudCtx.deferredCtx.addDeferred(zedcloudCtx, key, buf, size, url, bailOnHTTPErr)
+	zedcloudCtx.deferredCtx.addDeferred(zedcloudCtx, key, buf, size, url, bailOnHTTPErr, itemType)
 }
 
 func (ctx *DeferredContext) addDeferred(zedcloudCtx *ZedCloudContext,
-	key string, buf *bytes.Buffer, size int64, url string, bailOnHTTPErr bool) {
+	key string, buf *bytes.Buffer, size int64, url string, bailOnHTTPErr bool, itemType interface{}) {
+	ctx.lock.Lock()
+	defer ctx.lock.Unlock()
 
 	log := zedcloudCtx.log
-	log.Functionf("AddDeferred(%s) size %d map %d\n", key,
+	log.Functionf("AddDeferred(%s) size %d items %d\n", key,
 		size, len(ctx.deferredItems))
 	if len(ctx.deferredItems) == 0 {
 		startTimer(log, ctx)
 	}
-	l, ok := ctx.deferredItems[key]
-	if ok {
-		log.Tracef("Appending to key %s have %d\n", key, len(l.list))
-	} else {
-		log.Tracef("Adding key %s\n", key)
+	found := false
+	ind := 0
+	var itemList deferredItemList
+	for ind, itemList = range ctx.deferredItems {
+		if itemList.key == key {
+			found = true
+			break
+		}
 	}
 	item := deferredItem{
+		added:         time.Now(),
 		buf:           buf,
 		size:          size,
 		url:           url,
 		zedcloudCtx:   zedcloudCtx,
 		bailOnHTTPErr: bailOnHTTPErr,
 	}
-	l.list = append(l.list, item)
-	ctx.deferredItems[key] = l
+	if found {
+		log.Tracef("Appending to key %s have %d\n", key, len(itemList.list))
+		itemList.list = append(itemList.list, item)
+		ctx.deferredItems[ind] = itemList
+	} else {
+		log.Tracef("Adding key %s\n", key)
+		l := deferredItemList{key: key, itemType: itemType}
+		l.list = append(l.list, item)
+		ctx.deferredItems = append(ctx.deferredItems, l)
+	}
+}
+
+//TriggerDeferredTimer trigger deferred event immediately
```

Comment:
Why do you want to do this immediately after a failure? Triggering a minute or 10 seconds later might make sense  but not immediately AFAIU

Reply:
Not after a failure but after the success

Reply:
I remove this one and use one item send per adding to deferred